crates/oxc_minifier/src/ast_passes/peephole_replace_known_methods/mod.rs

```diff
@@ -41,25 +41,158 @@ impl PeepholeReplaceKnownMethods {
-                        ctx.ast.string_literal(call_expr.span, string_lit.value.cow_to_lowercase()),
-                    ),
-                    "toUpperCase" => Some(
-                        ctx.ast.string_literal(call_expr.span, string_lit.value.cow_to_uppercase()),
-                    ),
-                    "trim" => Some(ctx.ast.string_literal(call_expr.span, string_lit.value.trim())),
-                    _ => None,
-                };
-
-                if let Some(transformed_value) = transformed_value {
-                    self.changed = true;
-                    *node = ctx.ast.expression_from_string_literal(transformed_value);
+            match call_expr.arguments.len() {
+                0 => {
+                    let transformed_value =
+                        match member.property.name.as_str() {
+                            "toLowerCase" => Some(ctx.ast.string_literal(
+                                call_expr.span,
+                                string_lit.value.cow_to_lowercase(),
+                            )),
+                            "toUpperCase" => Some(ctx.ast.string_literal(
+                                call_expr.span,
+                                string_lit.value.cow_to_uppercase(),
+                            )),
+                            "trim" => Some(
+                                ctx.ast.string_literal(call_expr.span, string_lit.value.trim()),
+                            ),
+                            _ => None,
+                        };
+
+                    if let Some(transformed_value) = transformed_value {
+                        self.changed = true;
+                        *node = ctx.ast.expression_from_string_literal(transformed_value);
+                    }
                 }
+                _ => match member.property.name.as_str() {
+                    "indexOf" | "lastIndexOf" => {
+                        self.try_fold_index_of(node, ctx);
+                    }
+                    "substr" => {
+                        self.try_fold_string_substr(node, ctx);
+                    }
+                    "substring" | "slice" => self.try_fold_string_substring_or_slice(node, ctx),
+                    "charAt" => self.try_fold_string_char_at(node, ctx),
+                    "charCodeAt" => self.try_fold_string_char_code_at(node, ctx),
+                    "replace" => self.try_fold_string_replace(node, ctx),
+                    "replaceAll" => self.try_fold_string_replace_all(node, ctx),
+                    _ => {}
+                },
             }
         }
     }
+
+    fn try_fold_index_of<'a>(&mut self, node: &mut Expression<'a>, ctx: &mut TraverseCtx<'a>) {
+        let Expression::CallExpression(call_expr) = node else { unreachable!() };
+        let Expression::StaticMemberExpression(member) = &call_expr.callee else { unreachable!() };
+        let Expression::StringLiteral(string_lit) = &member.object else { unreachable!() };
```

Comment:
i couldn't find a decent way to pass these in via args to this function without wanting to kill the borrow checker. if anyone has an idea for a better way please let me know 🙂

Reply:
You need to pass in immutable parameters and return an `Option<Expression<'a>>`, then make the *expr assignment outside of this function.

Take a look at `peephole_fold_constants`, they follow this pattern for all `fold` methods.